lib/pm_lib.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * PM Echo - Path Monitoring lib functions
+ * Copyright (C) 6WIND 2019
+ *
+ * This file is part of FRR.
+ *
+ * FRR is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2, or (at your option) any
+ * later version.
+ *
+ * FRR is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#include "command.h"
+#include "memory.h"
+#include "prefix.h"
+#include "thread.h"
+#include "stream.h"
+#include "zclient.h"
+#include "table.h"
+#include "vty.h"
+#include "pm_lib.h"
+
+DEFINE_MTYPE_STATIC(LIB, PM_INFO, "PM info")
+
+int pm_debug = 0;
+struct pm_gbl pm_gbl;
+
+/*
+ * pm_gbl_init - Initialize the PM global structure
+ */
+void pm_gbl_init(void)
```

Comment:
Please don't use this code.

I don't mean to insult anyone but I think this could be better written, here are some points against it:

- It assumes your daemon will keep feeding its integration protocols information (e.g. settings change, counters etc...). This is actually one problem we have with BFD because we expect it to work with `ptm-bfd` (which has no `show` commands) and `bfdd` (see issue #4485 ).

- The logic of settings management is a bit convoluted: it expect to own all your session settings in the protocol integration code (not your daemon). You'll have synchronization issues with your daemon later.

- The message building code is awful because if you don't have a information it will omit from the packet, then you have to do all kinds of logic in your daemon to support readings messages from it.

   Here is the BFD message format for reference:

   ```
	/*
	 * Register/Deregister/Update Message format:
	 * - header: Command, VRF
	 * - l: pid
	 * - w: family
	 *   - AF_INET:
	 *     - l: destination ipv4
	 *   - AF_INET6:
	 *     - 16 bytes: destination IPv6
	 * - command != ZEBRA_BFD_DEST_DEREGISTER
	 *   - l: min_rx
	 *   - l: min_tx
	 *   - c: detect multiplier
	 * - c: is_multihop?
	 *   - multihop:
	 *     - w: family
	 *       - AF_INET:
	 *         - l: destination ipv4
	 *       - AF_INET6:
	 *         - 16 bytes: destination IPv6
	 *     - c: ttl
	 *   - no multihop
	 *     - AF_INET6:
	 *       - w: family
	 *       - 16 bytes: ipv6 address
	 *     - c: ifname length
	 *     - X bytes: interface name
	 * - c: bfd_cbit
	 *
	 * q(64), l(32), w(16), c(8)
	 */
   ```

   Example:
   - If command is not `DEREGISTER` then the message has timers.
   - If you have a `single-hop` peer with `IPv6` then it expects a `source` address. It could be made simplier by ALWAYS including the source address (even if blank).